// @node-red/runtime/lib/projects/git/index.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var fs = require("fs-extra");
var fspath = require("path");

var gitTools = require("./git");
var sshTools = require("./ssh");
var Projects = require("./Project");

var settings;
var runtime;
var layout;
var projectsEnabled = false;
var projectsDir;
var globalGitUser = false;
var activeProject = null;

// Any condition that prevents loading flows from a project: an empty repo, a
// missing package.json, or an in-progress merge. Returned as a {code,message}
// descriptor (or null) so the runtime owns turning it into a thrown error.
function getProjectError(project) {
    if (project.isEmpty()) {
        return {
            code: "project_empty",
            message: "Project repository is empty"
        };
    }
    if (project.missingFiles && project.missingFiles.indexOf("package.json") !== -1) {
        return {
            code: "missing_package_file",
            message: "Project missing package.json"
        };
    }
    if (project.isMerging()) {
        return {
            code: "git_merge_conflict",
            message: "Project has unmerged changes. Cannot load flows"
        };
    }
    return null;
}

function init(_settings, _runtime, _layout) {
    settings = _settings;
    runtime = _runtime;
    layout = _layout;
    projectsEnabled = false;
    projectsDir = fspath.resolve(fspath.join(settings.userDir,"projects"));
    activeProject = null;
    var projectSettings = ((settings.editorTheme || {}).projects) || {};

    try {
        if (projectSettings.enabled === true) {
            projectsEnabled = true;
        } else if (projectSettings.enabled === false) {
            runtime.log.warn(runtime.log._("storage.localfilesystem.projects.disabled"));
        }
    } catch(err) {
        runtime.log.warn(runtime.log._("storage.localfilesystem.projects.disabledNoFlag"));
        projectsEnabled = false;
    }

    if (projectSettings.path) {
        projectsDir = projectSettings.path;
    }

    if (!projectsEnabled) {
        return Promise.resolve();
    }

    return sshTools.init(settings,runtime).then(function() {
        return gitTools.init(settings).then(function(gitConfig) {
            if (!gitConfig || /^1\./.test(gitConfig.version)) {
                if (!gitConfig) {
                    runtime.log.warn(runtime.log._("storage.localfilesystem.projects.git-not-found"));
                } else {
                    runtime.log.warn(runtime.log._("storage.localfilesystem.projects.git-version-old",{version:gitConfig.version}));
                }
                projectsEnabled = false;
                if (settings.editorTheme && settings.editorTheme.projects) {
                    settings.editorTheme.projects.enabled = false;
                }
                return;
            }
            settings.editorTheme.projects.workflow = {
                mode: (settings.editorTheme.projects.workflow || {}).mode || "manual"
            };
            globalGitUser = gitConfig.user;
            Projects.init(settings,runtime,layout);
            if (!settings.readOnly) {
                return fs.ensureDir(projectsDir);
            }
        });
    });
}

function getProjectPath(name) {
    var projectPath = fspath.resolve(fspath.join(projectsDir,name));
    var relativePath = fspath.relative(projectsDir,projectPath);
    if (/^\.\./.test(relativePath)) {
        throw new Error("Invalid project name");
    }
    return projectPath;
}

async function setLegacyCredentialSecret(project, projectName) {
    // Older installs stored one top-level credentialSecret rather than a
    // per-project entry in the persisted projects settings. When such a
    // project is first loaded, copy that legacy value into the project's
    // settings record and mirror it onto the loaded project so subsequent
    // project operations use the canonical per-project secret.
    if (project.credentialSecret === undefined && settings.credentialSecret !== undefined) {
        var globalProjectSettings = (settings.get && settings.get("projects")) || {};
        globalProjectSettings.projects = globalProjectSettings.projects || {};
        globalProjectSettings.projects[projectName] = globalProjectSettings.projects[projectName] || {};
        if (globalProjectSettings.projects[projectName].credentialSecret !== settings.credentialSecret) {
            globalProjectSettings.projects[projectName].credentialSecret = settings.credentialSecret;
            await settings.set("projects", globalProjectSettings);
        }
        project.credentialSecret = settings.credentialSecret;
    }
    return project;
}

async function loadProject(projectName) {
    var project = await Projects.load(getProjectPath(projectName));
    await setLegacyCredentialSecret(project, projectName);
    return project;
}

module.exports = {
    init: init,
    available: function() {
        return projectsEnabled;
    },
    getProjectsDir: function() {
        return projectsDir;
    },
    getGlobalGitUser: function() {
        return globalGitUser;
    },
    setActiveProject: function(project) {
        activeProject = project || null;
    },
    flowFileExists: function() {
        return layout.flowFileExists ? layout.flowFileExists() : true;
    },
    exportFilesForEditor: function() {
        return layout.exportFilesForEditor();
    },
    // What blocks loading flows from this project (any project error), or null.
    getLoadError: function(project) {
        return getProjectError(project);
    },
    // What blocks saving flows to this project, or null. Only a merge conflict
    // blocks a write — an empty or package-less project can still be written to
    // (that's how it gets populated).
    getSaveError: function(project) {
        var error = getProjectError(project);
        if (error && error.code === "git_merge_conflict") {
            return error;
        }
        return null;
    },
    loadInitialProject: async function(projectName) {
        if (!projectsEnabled) {
            return null;
        }
        if (!projectName) {
            // Legacy boot fallback for older project installs that predate persisted
            // projects.activeProject in .config.projects.json. If no active project is configured yet,
            // treat the old top-level settings.flowFile value as a project name, load it if possible,
            // and let runtime/projects persist activeProject so future boots use the
            // canonical setting instead.
            if (settings.flowFile) {
                try {
                    return await loadProject(settings.flowFile);
                } catch(err) {
                    return null;
                }
            }
            return null;
        }
        return loadProject(projectName);
    },
    listProjects: function() {
        return fs.readdir(projectsDir).then(function(fns) {
            var dirs = [];
            fns.sort(function(A,B) {
                return A.toLowerCase().localeCompare(B.toLowerCase());
            }).forEach(function(fn) {
                var fullPath = fspath.join(projectsDir,fn);
                if (fn[0] !== ".") {
                    var stats = fs.lstatSync(fullPath);
                    if (stats.isDirectory()) {
                        dirs.push(fn);
                    }
                }
            });
            return dirs;
        });
    },
    loadProject: loadProject,
    createProject: function(user, metadata) {
        metadata.path = getProjectPath(metadata.name);
        return Projects.create(user, metadata);
    },
    deleteProject: function(user, projectName) {
        return Projects.delete(user, getProjectPath(projectName));
    },
    initialiseProject: function(project, user, data) {
        return project.initialise(user,data);
    },
    updateProject: function(project, user, data) {
        return project.update(user,data);
    },
    ssh: sshTools
};
