// @node-red/runtime/lib/projects/index.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var runtime;
var activeProjectType;
var activeProject;
var fspath = require("path");

// Registry of project types. A storage module selects one by name via
// storage.getProjectType() and supplies the flow layout it should use. New
// types (e.g. a non-git versioning backend) can register here.
var projectTypes = {
    git: require("./git")
};

function isObject(value) {
    return !!value && typeof value === "object" && !Array.isArray(value);
}

function normaliseProjectSettings(settings) {
    var normalised = isObject(settings) ? Object.assign({}, settings) : {};
    if (!isObject(normalised.projects)) {
        normalised.projects = {};
    }
    return normalised;
}

// Resolve the project type selected by the active storage module: it declares
// a project type + the flow layout to drive it with (e.g.
// localfilesystem -> "git" + its single-file layout).
async function resolveActiveProjectType() {
    var typeName = runtime.storage.getProjectType && runtime.storage.getProjectType();
    if (typeName && projectTypes[typeName]) {
        var projectType = projectTypes[typeName];
        var flowLayout = runtime.storage.getFlowLayout ? runtime.storage.getFlowLayout() : null;
        await projectType.init(runtime.settings, runtime, flowLayout);
        return projectType;
    }
    return null;
}

function reloadActiveProject(action, clearContext) {
    return runtime.nodes.stopFlows().then(function() {
        if (clearContext) {
            return runtime.nodes.clearContext();
        }
    }).then(function() {
        return runtime.nodes.loadFlows(true).then(function() {
            runtime.events.emit("runtime-event",{id:"project-update", payload:{ project: activeProject.name, action:action}});
        }).catch(function(err) {
            runtime.events.emit("runtime-event",{id:"project-update", payload:{ project: activeProject.name, action:action}});
            throw err;
        });
    }).catch(function(err) {
        console.log(err.stack);
        throw err;
    });
}

function ensureActiveProject(projectName) {
    if (!activeProject || activeProject.name !== projectName) {
        throw new Error("Cannot operate on inactive project wanted:"+projectName+" current:"+(activeProject&&activeProject.name));
    }
}

function getAutoWorkflowMode(user) {
    var username = user && user.username ? user.username : "_";
    var userSettings = runtime.settings.getUserSettings(username) || {};
    var gitSettings = userSettings.git || {};
    return (gitSettings.workflow||{}).mode || runtime.settings.editorTheme.projects.workflow.mode;
}

async function maybeAutoCommit(user) {
    if (getAutoWorkflowMode(user) === "auto") {
        await activeProject.stageFile(activeProject.listAutosaveFiles());
        var result = await activeProject.status(user, false);
        var items = Object.values(result.files || {});
        if (items.findIndex((item) => (item.status === "M ")) >= 0) {
            return activeProject.commit(user,{message:"Update flow files"});
        }
    }
}

// Turn a project type's {code,message} error descriptor into a thrown Error.
// No-op when there is no error.
function throwProjectError(projectError) {
    if (projectError) {
        var error = new Error(projectError.message);
        error.code = projectError.code;
        throw error;
    }
}

module.exports = {
    // Register a project type implementation under a name a storage module can
    // select via getProjectType(). "git" is registered by default.
    registerProjectType: function(name, impl) {
        projectTypes[name] = impl;
    },
    init: async function(_runtime) {
        runtime = _runtime;
        activeProjectType = await resolveActiveProjectType();
        activeProject = null;

        if (!activeProjectType || !activeProjectType.available || !activeProjectType.available()) {
            return;
        }

        var rawSettings = runtime.settings.get("projects");
        var saveSettings = !isObject(rawSettings) || !isObject(rawSettings.projects);
        var globalSettings = normaliseProjectSettings(rawSettings);
        var configuredProject = globalSettings.activeProject;
        activeProject = await activeProjectType.loadInitialProject(configuredProject);
        activeProjectType.setActiveProject(activeProject);
        if (activeProject && configuredProject !== activeProject.name) {
            globalSettings.activeProject = activeProject.name;
            saveSettings = true;
        }
        if (saveSettings) {
            await runtime.settings.set("projects", globalSettings);
        }
        if (activeProject) {
            if (activeProjectType.getProjectsDir) {
                runtime.log.info(runtime.log._("storage.localfilesystem.projects.projects-directory", {projectsDirectory: activeProjectType.getProjectsDir()}));
            }
            runtime.log.info(runtime.log._("storage.localfilesystem.projects.active-project",{project:activeProject.name||"none"}));
        } else {
            runtime.log.warn(runtime.log._("storage.localfilesystem.projects.no-active-project"));
        }
    },
    // Whether the projects feature is available with the current storage module and settings. 
    available: function() {
        return activeProjectType ? activeProjectType.available() : false;
    },
    getActiveProject: function() {
        return activeProject;
    },
    // Throw if the active project can't currently serve a flow load (empty repo,
    // missing package.json, merge conflict). No-op when no project is active.
    // The project type owns the error codes/policy; we own throwing them.
    assertCanLoadFlows: function() {
        if (!activeProjectType || !activeProject) {
            return;
        }
        throwProjectError(activeProjectType?.getLoadError?.(activeProject));
    },
    // Throw if the active project can't currently accept a flow save (merge
    // conflict only). No-op when no project is active.
    assertCanSaveFlows: function() {
        if (!activeProjectType || !activeProject) {
            return;
        }
        throwProjectError(activeProjectType?.getSaveError?.(activeProject));
    },
    // Whether the active flow file exists on disk. Gates whether RED.settings.files
    // is advertised when there is no active project.
    // If method missing from project type, assume it exists (e.g. for a non-file-based project type).
    flowFileExists: function() {
        return activeProjectType ? activeProjectType.flowFileExists() : false;
    },
    // The default { flow, credentials } file names for RED.settings.files. Only
    // used when no project is active (see api/settings.js): the project type
    // returns the resolved default paths, which are absolute (userDir + name).
    // The editor uses these to pre-fill the flow/credentials names when creating
    // a project, so it wants bare file names to prepopulate the UI fields.
    getDefaultProjectFilesForEditor: function() {
        var files = activeProjectType ? activeProjectType.exportFilesForEditor() : {};
        return {
            flow: files.flow ? fspath.basename(files.flow) : files.flow,
            credentials: files.credentials ? fspath.basename(files.credentials) : files.credentials
        };
    },
    getGlobalGitUser: function() {
        return activeProjectType?.getGlobalGitUser?.() ?? false;
    },
    ssh: {
        listSSHKeys: function() {
            return activeProjectType?.ssh?.listSSHKeys?.() ?? [];
        },
        getSSHKey: function() {
            return activeProjectType?.ssh?.getSSHKey?.() ?? null;
        },
        generateSSHKey: function() {
            return activeProjectType?.ssh?.generateSSHKey?.() ?? null;
        },
        deleteSSHKey: function() {
            return activeProjectType?.ssh?.deleteSSHKey?.() ?? null;
        }
    },
    listProjects: function() {
        return activeProjectType?.listProjects?.() ?? [];
    },
    createProject: async function(user, metadata) {
        // Migrating an existing install into its first project requires no
        // active project. The selected project type owns how files are migrated.
        if (metadata.files && metadata.migrateFiles && activeProject) {
            throw new Error("Cannot migrate as there is an active project");
        }
        if (metadata.files && metadata.migrateFiles) {
            // Re-export the live credentials so the storage backend can write
            // them into the new project. Resolving the key and exporting the
            // running credentials is the runtime's job, not the backend's.
            var currentEncryptionKey = runtime.settings.get("credentialSecret");
            if (currentEncryptionKey === undefined) {
                currentEncryptionKey = runtime.settings.get("_credentialSecret");
            }
            if (!metadata.hasOwnProperty("credentialSecret")) {
                metadata.credentialSecret = currentEncryptionKey;
            }
            // Re-key the runtime to the new project's secret BEFORE exporting, so
            // the exported credentials are encrypted with that key. exportCredentials
            // returns the already-encrypted blob (current key) when not dirty, so
            // without this a migration that sets a different credentialSecret would
            // write a blob the new project can't decrypt on first load.
            runtime.nodes.setCredentialSecret(metadata.credentialSecret);
            metadata.migratedCredentials = await runtime.nodes.exportCredentials();
        }
        var createdProject = await activeProjectType.createProject(user, metadata);
        await module.exports.setActiveProject(user, createdProject.name);
        return module.exports.getProject(user, createdProject.name);
    },
    setActiveProject: async function(user, projectName, clearContext) {
        activeProject = await activeProjectType.loadProject(projectName);
        activeProjectType.setActiveProject(activeProject);
        var globalSettings = normaliseProjectSettings(runtime.settings.get("projects"));
        globalSettings.activeProject = activeProject.name;
        await runtime.settings.set("projects", globalSettings);
        runtime.log.info(runtime.log._("storage.localfilesystem.projects.changing-project",{project:(activeProject&&activeProject.name)||"none"}));
        return reloadActiveProject("loaded", clearContext);
    },
    getProject: async function(user, name) {
        ensureActiveProject(name);
        activeProject = await activeProjectType.loadProject(name);
        activeProjectType.setActiveProject(activeProject);
        return activeProject.export();
    },
    deleteProject: async function(user, name) {
        if (activeProject && activeProject.name === name) {
            var e = new Error("NLS: Can't delete the active project");
            e.code = "cannot_delete_active_project";
            throw e;
        }
        return activeProjectType.deleteProject(user, name);
    },
    initialiseProject: async function(user, projectName, data) {
        ensureActiveProject(projectName);
        activeProject = await activeProjectType.initialiseProject(activeProject, user, data);
        activeProjectType.setActiveProject(activeProject);
        runtime.nodes.setCredentialSecret(activeProject.credentialSecret);
        return reloadActiveProject("updated");
    },
    updateProject: async function(user, projectName, data) {
        ensureActiveProject(projectName);
        var isReset = data.resetCredentialSecret;
        var wasInvalid = activeProject.credentialSecretInvalid;
        var result = await activeProjectType.updateProject(activeProject, user, data);
        if (result.flowFilesChanged) {
            return reloadActiveProject("updated");
        } else if (result.credentialSecretChanged) {
            if (isReset || !wasInvalid) {
                if (isReset) {
                    runtime.nodes.clearCredentials();
                }
                runtime.nodes.setCredentialSecret(activeProject.credentialSecret);
                await runtime.nodes.exportCredentials().then(runtime.storage.saveCredentials);
                if (wasInvalid) {
                    return reloadActiveProject("updated");
                }
            } else if (wasInvalid) {
                return reloadActiveProject("updated");
            }
        }
    },
    handleFlowFileSave: async function(user) {
        if (!activeProject) {
            return;
        }
        await maybeAutoCommit(user);
    },
    getFiles: function(user, projectName) {
        ensureActiveProject(projectName);
        return activeProject.getFiles();
    },
    stageFile: function(user, projectName, file) {
        ensureActiveProject(projectName);
        return activeProject.stageFile(file);
    },
    unstageFile: function(user, projectName, file) {
        ensureActiveProject(projectName);
        return activeProject.unstageFile(file);
    },
    commit: function(user, projectName, options) {
        ensureActiveProject(projectName);
        var isMerging = activeProject.isMerging();
        return activeProject.commit(user, options).then(() => {
            if (isMerging && !activeProject.isMerging()) {
                return reloadActiveProject("merge-complete");
            }
        });
    },
    getFileDiff: function(user, projectName, file, type) {
        ensureActiveProject(projectName);
        return activeProject.getFileDiff(file,type);
    },
    getCommits: function(user, projectName, options) {
        ensureActiveProject(projectName);
        return activeProject.getCommits(options);
    },
    getCommit: function(user, projectName, sha) {
        ensureActiveProject(projectName);
        return activeProject.getCommit(sha);
    },
    getFile: function(user, projectName, filePath, sha) {
        ensureActiveProject(projectName);
        return activeProject.getFile(filePath,sha);
    },
    revertFile: function(user, projectName, filePath) {
        ensureActiveProject(projectName);
        return activeProject.revertFile(filePath).then(function() {
            return reloadActiveProject("revert");
        });
    },
    push: function(user, projectName, remoteBranchName, setRemote) {
        ensureActiveProject(projectName);
        return activeProject.push(user,remoteBranchName,setRemote);
    },
    pull: function(user, projectName, remoteBranchName, setRemote, allowUnrelatedHistories) {
        ensureActiveProject(projectName);
        return activeProject.pull(user,remoteBranchName,setRemote,allowUnrelatedHistories).then(function() {
            return reloadActiveProject("pull");
        });
    },
    getStatus: function(user, projectName, includeRemote) {
        ensureActiveProject(projectName);
        return activeProject.status(user, includeRemote);
    },
    resolveMerge: function(user, projectName, file, resolution) {
        ensureActiveProject(projectName);
        return activeProject.resolveMerge(file,resolution);
    },
    abortMerge: function(user, projectName) {
        ensureActiveProject(projectName);
        return activeProject.abortMerge().then(function() {
            return reloadActiveProject("merge-abort");
        });
    },
    getBranches: function(user, projectName, isRemote) {
        ensureActiveProject(projectName);
        return activeProject.getBranches(user, isRemote);
    },
    deleteBranch: function(user, projectName, branch, isRemote, force) {
        ensureActiveProject(projectName);
        return activeProject.deleteBranch(user, branch, isRemote, force);
    },
    setBranch: function(user, projectName, branchName, isCreate) {
        ensureActiveProject(projectName);
        return activeProject.setBranch(branchName,isCreate).then(function() {
            return reloadActiveProject("change-branch");
        });
    },
    getBranchStatus: function(user, projectName, branchName) {
        ensureActiveProject(projectName);
        return activeProject.getBranchStatus(branchName);
    },
    getRemotes: function(user, projectName) {
        ensureActiveProject(projectName);
        return activeProject.getRemotes(user);
    },
    addRemote: function(user, projectName, options) {
        ensureActiveProject(projectName);
        return activeProject.addRemote(user, options.name, options);
    },
    removeRemote: function(user, projectName, remote) {
        ensureActiveProject(projectName);
        return activeProject.removeRemote(user, remote);
    },
    updateRemote: function(user, projectName, remote, body) {
        ensureActiveProject(projectName);
        return activeProject.updateRemote(user, remote, body);
    }
};
