// @node-red/runtime/lib/storage/localfilesystem/flows.js
/**
 * Copyright JS Foundation and other contributors, http://js.foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

var fs = require("fs-extra");
var os = require("os");
var fspath = require("path");

var log = require("@node-red/util").log;
var util = require("./util");

var settings;
var runtime;
var initialFlowLoadComplete = false;
var flowsFileExists = false;
var usingHostName = false;

var defaultContext;

// --- path format utilities ---------------------------------------------------------

// Filename conventions for the single-file flow layout: the credentials file
// derives a "_cred" suffix from the flow file, and a backup is a ".<name>.backup"
// sibling. These are this layout's naming rules, not generic storage utilities.
function getBackupFilename(filename) {
    filename = filename || "undefined";
    var ffName = fspath.basename(filename);
    var ffDir = fspath.dirname(filename);
    return fspath.join(ffDir, "." + ffName + ".backup");
}

function getCredentialsFilename(filename) {
    filename = filename || "undefined";
    var ffDir = fspath.dirname(filename);
    var ffExt = fspath.extname(filename);
    var ffBase = fspath.basename(filename, ffExt);
    return fspath.join(ffDir, ffBase + "_cred" + ffExt);
}

function getDefaultFlowFilename() {
    var flowFile = settings.flowFile;
    usingHostName = false;

    if (!flowFile) {
        flowFile = "flows_"+os.hostname()+".json";
        usingHostName = true;
        return fspath.join(settings.userDir, flowFile);
    }
    if (fspath.isAbsolute(flowFile)) {
        return flowFile;
    }
    if (flowFile.substring(0,2) === "./") {
        return fspath.join(process.cwd(), flowFile);
    }
    var cwdPath = fspath.join(process.cwd(), flowFile);
    try {
        fs.statSync(cwdPath);
        return cwdPath;
    } catch(err) {
        return fspath.join(settings.userDir, flowFile);
    }
}

function getDefaultCredentialsFilename(flowFile) {
    var ffExt = fspath.extname(flowFile);
    var ffBase = fspath.basename(flowFile, ffExt);
    var credentialsFile = fspath.join(settings.userDir, ffBase+"_cred"+ffExt);
    if (!fs.existsSync(credentialsFile)) {
        credentialsFile = getCredentialsFilename(flowFile);
    }
    return credentialsFile;
}

// --- project path utilities ---------------------------------------------------------

function isPathInsideProject(projectPath, filePath) {
    var relativePath = fspath.relative(projectPath, filePath);
    return relativePath && !/^\.\./.test(relativePath) && !fspath.isAbsolute(relativePath);
}

function toManifestPath(project, filePath) {
    if (!filePath) {
        return filePath;
    }
    var projectRoot = (project.paths && project.paths.root) || "";
    if (/^\.\./.test(fspath.relative(project.path, fspath.join(project.path, filePath)))) {
        throw new Error("Invalid file path: " + filePath);
    }
    return filePath.substring(projectRoot.length);
}

// --- project file utilities ---------------------------------------------------------

function withProjectPathRoot(projectPath, files) {
    return {
        flowFile: files?.flowFile ? fspath.join(projectPath, files.flowFile) : undefined,
        credentialsFile: files?.credentialsFile ? fspath.join(projectPath, files.credentialsFile) : undefined
    }
}

// This defines the baseline file targets the runtime should use
// when projects are disabled or no project is active.
function getDefaultFiles() {
    var flowFile = getDefaultFlowFilename();
    var credentialsFile = getDefaultCredentialsFilename(flowFile);
    return {
        flowFile: flowFile,
        credentialsFile: credentialsFile,
    };
}

// Resolve the flow & credentials file references recorded in a project's
// package metadata. Returns the manifest-declared paths relative to the
// project root.
function getProjectManifestFiles(project) {
    var nrSettings = project.package?.["node-red"]?.settings || {};
    return withProjectPathRoot(project.paths.root, nrSettings);
}

// Context selection is separate because the runtime has to choose between
// default storage and project-backed storage on every read/write boundary. The
// project case is intentionally not treated as a normal overlay: if a project
// resolves no flow files, we must fail in project scope rather than silently
// falling back to the global default files.
function getProjectFiles(project) {
    if (!project) {
        return {...defaultContext};
    }
    var files = getProjectManifestFiles(project);
    return withProjectPathRoot(project.path, files);
}

// --- storage context management ---------------------------------------------------------

// The project backing the current read/write, or null when projects are
// disabled / none is active. Read straight from the runtime each call so flows
// never caches a stale active project.
function getActiveProject() {
    return runtime.projects ? runtime.projects.getActiveProject() : null;
}

// --- project creation utilities ---------------------------------------------------------

// Resolve the absolute flow + credentials paths for a project being created or
// migrated, rejecting anything that escapes the project directory. Callers
// (createFiles / migrateFiles) default the file names before calling this; here
// we only join + validate. Shared by project.createFiles / project.migrateFiles.
function getProjectCreationFiles(projectPath, files) {
    var flowFile = fspath.join(projectPath, files.flow);
    if (!isPathInsideProject(projectPath, flowFile)) {
        throw new Error("Flow file path is outside the project directory");
    }
    var credentialsFile = fspath.join(projectPath, files.credentials);
    if (!isPathInsideProject(projectPath, credentialsFile)) {
        throw new Error("Credentials file path is outside the project directory");
    }
    return { 
        flowFile: flowFile, 
        credentialsFile: credentialsFile 
    };
}

// --- storage module interface -------------------------------------------------

function getPrettyFlag(project) {
    return settings.flowFilePretty || (project && settings.flowFilePretty !== false);
}

async function getFlows() {
    var project = getActiveProject();
    var context = getProjectFiles(project);
    if (!initialFlowLoadComplete) {
        initialFlowLoadComplete = true;
        log.info(log._("storage.localfilesystem.user-dir",{path:settings.userDir}));
        if (usingHostName && !project) {
            log.warn(log._("storage.localfilesystem.warn_name"));
        }
        log.info(log._("storage.localfilesystem.flows-file",{path:context.flowFile}));
    }
    // active project but no flow file, so fail in project scope rather than
    // silently reading/writing the global default file.
    if (project && !context.flowFile) {
        var missingFlowFileError = new Error("Project has no flow file");
        missingFlowFileError.code = "missing_flow_file";
        throw missingFlowFileError;
    }
    return util.readFile(context.flowFile,getBackupFilename(context.flowFile),null,"flow").then(function(result) {
        if (result === null) {
            flowsFileExists = false;
            return [];
        }
        flowsFileExists = true;
        return result;
    });
}

async function saveFlows(flows) {
    if (settings.readOnly) {
        return;
    }
    var project = getActiveProject();
    var context = getProjectFiles(project);

    flowsFileExists = true;
    var flowData;
    if (getPrettyFlag(project)) {
        flowData = JSON.stringify(flows,null,4);
    } else {
        flowData = JSON.stringify(flows);
    }
    return util.writeFile(context.flowFile, flowData, getBackupFilename(context.flowFile));
}

function getCredentials() {
    var project = getActiveProject();
    var context = getProjectFiles(project);
    return util.readFile(context.credentialsFile, getBackupFilename(context.credentialsFile), {}, 'credentials');
}

async function saveCredentials(credentials) {
    if (settings.readOnly) {
        return;
    }
    var project = getActiveProject();
    var context = getProjectFiles(project);
    var credentialData;
    if (getPrettyFlag(project)) {
        credentialData = JSON.stringify(credentials,null,4);
    } else {
        credentialData = JSON.stringify(credentials);
    }
    return util.writeFile(context.credentialsFile, credentialData, getBackupFilename(context.credentialsFile));
}

module.exports = {
    init: async function(_settings, _runtime) {
        settings = _settings;
        runtime = _runtime;
        initialFlowLoadComplete = false;
        flowsFileExists = false;
        defaultContext = getDefaultFiles();
    },
    getFlows: getFlows,
    saveFlows: saveFlows,
    getCredentials: getCredentials,
    saveCredentials: saveCredentials,

    // The flow layout the built-in git project type drives. This is the seam
    // between a storage module and the runtime projects service for project-aware
    // flow storage. Storage backends that support projects expose the same shape
    // via `getFlowLayout()`, while backends that do not support projects can
    // ignore it.
    //
    // Single-file flow storage and its project layout are one concern (shared
    // file-location model + runtime active-project resolution), so the layout
    // operations live here directly on `layout`.
    layout: {
        // --- inspect: report on an existing project's data files ---------

        // Whether the active flow file exists on disk yet. Disk state, tracked
        // as flows are read/written — not a config check (cf. flowsDeclared).
        flowFileExists: function() {
            return flowsFileExists;
        },
        // Whether a project declares any flow files in its package metadata. A
        // config check, not a disk-existence check (cf. flowFileExists).
        flowsDeclared: function(project) {
            var files = getProjectManifestFiles(project);
            return !!files.flowFile;
        },
        // The editor's per-project file projection (becomes Project.export().files,
        // and RED.settings.files when no project is active), read by name in the
        // project-settings panel.
        // @returns {{flow, credentials}} root-relative paths — a NAMED pair, vs
        //   the ARRAY from listAutosaveFiles. The git type adds package.json.
        exportFilesForEditor: function(project) {
            var files = project ? getProjectManifestFiles(project) : defaultContext;
            return {
                flow: files.flowFile,
                credentials: files.credentialsFile
            };
        },
        // The flow data files (flow + credentials) git should stage for a
        // project. Repo-level files (package.json, README.md) are owned by the
        // git type.
        // @returns {string[]} root-relative paths — an ARRAY because git stages
        //   a set of N files (a multi-file layout would return more); the same
        //   underlying data as exportFilesForEditor, shaped for `git add`.
        listAutosaveFiles: function(project) {
            var files = getProjectManifestFiles(project);
            return [files.flowFile, files.credentialsFile].filter(Boolean);
        },
        
        // --- provision: set up a new project's data files ----------------

        // Create the initial flow + credentials data files for a new project.
        // Defaults the flow name (and derives the credentials name) when the
        // request omits them, normalises project.files in place, and returns
        // the relative file names to stage plus a promise for the writes.
        createFiles: function(project) {
            var files = project.files || {};
            if (!files.flow) {
                files.flow = "flows.json";
            }
            if (!files.credentials) {
                files.credentials = getCredentialsFilename(files.flow);
            }
            project.files = files; // ensure the project has the normalised file manifest for later layout operations

            var paths = getProjectCreationFiles(project.path, project.files);
            return {
                files: [project.files.flow, project.files.credentials],
                write: Promise.all([
                    util.writeFile(paths.flowFile, "[]"),
                    util.writeFile(paths.credentialsFile, "{}")
                ])
            };
        },
        // Provision a new project by migrating the pre-project install: copies
        // the old flow file into the project and writes the migrated credentials.
        // The runtime exports/re-keys the live credentials and hands them over on
        // project.migratedCredentials, so write() is a plain promise like
        // createFiles (no credentials argument).
        migrateFiles: function(project) {
            var files = project.files || {};
            if (!files.flow) {
                files.flow = fspath.basename(defaultContext.flowFile);
            }
            if (!files.oldFlow) {
                files.oldFlow = defaultContext.flowFile;
            }
            if (!files.credentials) {
                files.credentials = fspath.basename(defaultContext.credentialsFile);
            }
            if (!files.oldCredentials) {
                files.oldCredentials = defaultContext.credentialsFile;
            }
            project.files = files; // ensure the project has the normalised file manifest for later layout operations

            var paths = getProjectCreationFiles(project.path, files);
            return {
                files: [files.flow, files.credentials],
                write: (function() {
                    var credentials = project.migratedCredentials;
                    var writes = [];
                    if (fs.existsSync(files.oldFlow)) {
                        writes.push(fs.copy(files.oldFlow, paths.flowFile));
                    } else {
                        writes.push(util.writeFile(paths.flowFile, "[]"));
                    }
                    var credentialData = settings.flowFilePretty ? JSON.stringify(credentials, null, 4) : JSON.stringify(credentials);
                    writes.push(util.writeFile(paths.credentialsFile, credentialData));
                    return Promise.all(writes);
                })()
            };
        },
        // Apply a flow/credentials rename to the project's package.json manifest.
        // Manifest-only by design: the data files are NOT moved on disk, so a
        // pointer to a not-yet-existing file yields an empty project until the
        // next save — matching the pre-project / upstream behaviour. Mutates
        // project.package and reports whether anything changed. The files/write 
        // result is provided for isomorphism with createFiles/migrateFiles, 
        // in case a future layout needs to provision or stage files on disk as part of a 
        // project-metadata update.
        updateFiles: function(project, files) {
            var changed = false;
            var nodeRedSettings = project.package["node-red"].settings;
            if (files.flow) {
                var manifestFlow = toManifestPath(project, files.flow);
                if (nodeRedSettings.flowFile !== manifestFlow) {
                    nodeRedSettings.flowFile = manifestFlow;
                    changed = true;
                }
            }
            if (files.credentials) {
                var manifestCredentials = toManifestPath(project, files.credentials);
                if (nodeRedSettings.credentialsFile !== manifestCredentials) {
                    nodeRedSettings.credentialsFile = manifestCredentials;
                    delete project.credentialSecretInvalid;
                    changed = true;
                }
            }
            return {
                changed: changed,

                // Shape-compatible with createFiles / migrateFiles so the caller
                // can treat every layout file operation the same way. The
                // single-file layout has nothing extra to stage or write here, but
                // a future layout could return real files / a write promise if needed.
                files: [],
                write: Promise.resolve()
            };
        },

        // --- manifest: the package.json record of where flows live -------

        // Record this layout's flow file references into a project's
        // package.json node-red settings. This is the single-file manifest.
        buildPackageSettings: function(project, nodeRedSettings, files) {
            nodeRedSettings.flowFile = toManifestPath(project, files.flow);
            nodeRedSettings.credentialsFile = toManifestPath(project, files.credentials);
        },

    }
};
