// @node-red/editor-client/src/tours/welcome-2-1.js
export default {
    steps: [
        {
            titleIcon: "fa fa-map-o",
            title: "Welcome to Node-RED 2.1!",
            description: "Let's take a moment to discover the new features in this release."
        },
        {
            title: "A new Tour Guide",
            description: "<p>First, as you've already found, we now have this tour of new features. We'll only show the tour the first time you open the editor for each new version of Node-RED.</p>"+
                         "<p>You can choose not to see this tour in the future by disabling it under the View tab of User Settings.</p>",
        },
        {
            prepare:function() {
                $("#red-ui-header-button-sidemenu").trigger("click");
                $("#menu-item-edit-menu").parent().addClass("open")
            },
            complete: function() {
                $("#menu-item-edit-menu").parent().removeClass("open")
            },
            element: "#menu-item-edit-menu-submenu",
            interactive: false,
            direction: "left",
            title: "New Edit menu",
            description: "<p>The main menu has been updated with a new 'Edit' section. This includes all of the familar options, like cut/paste and undo/redo.</p>"+
                         "<p>The menu now displays keyboard shortcuts for the options.</p>"

        },
        {
            prepare: function() {
                $("#red-ui-header-button-sidemenu").trigger("click");
                $("#menu-item-arrange-menu").parent().addClass("open")
            },
            complete: function() {
                $("#menu-item-arrange-menu").parent().removeClass("open")
            },
            element: "#menu-item-arrange-menu-submenu",
            interactive: false,
            direction: "left",
            title: "Arranging nodes",
            description: "<p>The new 'Arrange' section of the menu provides new options to help arrange your nodes. You can align them to a common edge, spread them out evenly or change their order.</p>",
        },
        {
            element: "#red-ui-workspace-tabs > li:first-child",
            title: "Flow and Group level environment variables",
            description: "<p>Flows and Groups can now have their own environment variables that can be referenced by nodes inside them.</p>",
        },
        {
            prepare: function(done) {
                RED.editor.editFlow(RED.nodes.workspace(RED.workspaces.active()),"editor-tab-envProperties");
                setTimeout(done,800);
            },
            element: "#red-ui-tab-editor-tab-envProperties-link-button",
            description: "<p>Flows and Groups now have an Environment Variables section in their edit dialog.</p>"
        },
        {
            element: ".node-input-env-container-row .red-ui-editableList-addButton",
            direction: "top",
            description: '<p>The environment variables are listed in this table and new ones can be added by clicking the <i class="fa fa-plus"></i> button.</p>',
            complete: function() {
                $("#node-dialog-cancel").trigger("click");
            }
        },
        {
            title: "And that's not all...",
            description: "<p>There's more still to come before 2.1.0 is released. Watch this space!</p>"
        },
    ]
}
