// @node-red/editor-client/src/tours/welcome.js
export default {
    version: "5.0.0",
    steps: [
        {
            titleIcon: "fa fa-map-o",
            title: {
                "en-US": "Welcome to Node-RED 5.0!",
                "ja": "Node-RED 5.0 へようこそ!",
                "fr": "Bienvenue dans Node-RED 5.0!"
            },
            description: {
                "en-US": "<p>Let's take a moment to discover the new features in this release.</p>",
                "ja": "<p>本リリースの新機能を見つけてみましょう。</p>",
                "fr": "<p>Prenons un moment pour découvrir les nouvelles fonctionnalités de cette version.</p>"
            }
        },
        {
            title: {
                "en-US": "A new editor experience"
            },
            description: {
                "en-US": `<p>This release brings a new design for the Node-RED editor, bringing a more modern look and feel, improved navigation and flexibility.</p>
                <p>Most notable are the sidebar panels which can be resized, collapsed, and customized to suit your workflow.</p>
                `
            }
        },
        {
            title: {
                "en-US": "Explorer sidebar"
            },
            description: {
                "en-US": `<p>The Information sidebar has been split into two. The explorer panel lets you easily browse your flows.</p>`
            },
            element: ".red-ui-info-outline",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-explorer-tab")
            }
        },
        {
            title: {
                "en-US": "Information sidebar"
            },
            description: {
                "en-US": `<p>The Information sidebar shows you details of whatever you have selected within the workspace. If the item has additional documentation included, it will be shown here.</p>`
            },
            element: ".red-ui-sidebar-info",
            direction: "left",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-info-tab")
            }
        }
        ,
        {
            title: {
                "en-US": "Node documentation icon"
            },
            description: {
                "en-US": `<p>Any nodes that have additional documentation will now show a docs icon. Clicking on the icon will show the documentation in a pop up.</p>`
            },
            image: 'images/editor-node-docs.png',
        }
        ,
        {
            title: {
                "en-US": "Palette sidebar"
            },
            description: {
                "en-US": `<p>The Palette is now a sidebar panel that can be moved to any of the other sidebar sections.</p>`
            },
            element: "#red-ui-palette-container",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-palette-tab")
            }
        },
        {
            title: {
                "en-US": "Sidebar toolbar"
            },
            description: {
                "en-US": `<p>All of the available sidebar panels can be accessed from the sidebar toolbar, including toggles to show/hide the entire sidebars</p>`
            },
            element: ".red-ui-sidebar-tab-bar > div",
            interactive: false
        },
        {
            title: {
                "en-US": "Dark Theme Unlocked"
            },
            description: {
                "en-US": `<p>Dark theme is now available to all users without needing to enable it in the settings.</p>
                          <p>This option may not be available if you have already installed a custom theme plugin.</p>`
            },
            element () {
                const themeSelector = $("#user-settings-view-dark-theme");
                if (themeSelector.length === 0) {
                    return $("#red-ui-tab-red-ui-settings-tab-view")
                }
                return themeSelector
            },
            interactive: false,
            prepare (done) {
                RED.actions.invoke('core:show-user-settings')
                 setTimeout(done,300);
            },
            complete () {
                $("#node-dialog-ok").trigger("click")
            }

        },
        {
            title: {
                "en-US": "Search toolbar"
            },
            description: {
                "en-US": `<p>The search toolbar allows you to quickly find nodes and flows within your workspace.</p>`
            },
            element: "#red-ui-view-searchtools-search",
            interactive: false,
            image: 'images/editor-search-toolbar.png',
        }
        ,
        {
            title: {
                "en-US": "Pausing debug"
            },
            description: {
                "en-US": `<p>The Debug sidebar allows you to pause its output; useful if you have a busy flow generating lots of messages.</p>`
            },
            prepare () {
                RED.actions.invoke("core:show-debug-tab")
            },
            element: "#red-ui-sidebar-debug-pause",
            interactive: false
        }
    ]
}
