// @node-red/editor-client/src/tours/welcome.js
export default {
    version: "5.0.0",
    steps: [
        {
            titleIcon: "fa fa-map-o",
            title: {
                "en-US": "Welcome to Node-RED 5.0!",
                "ja": "Node-RED 5.0 へようこそ!",
                "fr": "Bienvenue dans Node-RED 5.0!"
            },
            description: {
                "en-US": "<p>Let's take a moment to discover some of the new features in this release.</p>",
                "ja": "<p>本リリースの新機能を見つけてみましょう。</p>",
                "fr": "<p>Prenons un moment pour découvrir les nouvelles fonctionnalités de cette version.</p>"
            }
        },
        {
            title: {
                "en-US": "A new editor experience"
            },
            description: {
                "en-US": `<p>The Node-RED editor just got a redesign: cleaner, more flexible, more yours. Resize, collapse, and rearrange the sidebar panels to match your workflow.</p>
                `
            }
        },
        {
            title: {
                "en-US": "Explorer sidebar"
            },
            description: {
                "en-US": `<p>The Information sidebar is now two panels. Use the explorer panel to browse your flows.</p>`
            },
            element: ".red-ui-info-outline",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-explorer-tab")
            }
        },
        {
            title: {
                "en-US": "Information sidebar"
            },
            description: {
                "en-US": `<p>The Information sidebar shows details of whatever you select in the workspace. If an item includes documentation, you'll see it here.</p>`
            },
            element: ".red-ui-sidebar-info",
            direction: "left",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-info-tab")
            }
        }
        ,
        {
            title: {
                "en-US": "Node documentation icon"
            },
            description: {
                "en-US": `<p>Nodes with extra documentation now show a docs icon. Click it to open the docs in a pop-up.</p>`
            },
            image: 'images/editor-node-docs.png',
        }
        ,
        {
            title: {
                "en-US": "Palette sidebar"
            },
            description: {
                "en-US": `<p>The Palette is now a sidebar panel you can move to any other sidebar section.</p>`
            },
            element: "#red-ui-palette-container",
            interactive: false,
            prepare () {
                RED.actions.invoke("core:show-palette-tab")
            }
        },
        {
            title: {
                "en-US": "Sidebar toolbar"
            },
            description: {
                "en-US": `<p>Open any panel from the sidebar toolbar, including toggles to show or hide the sidebars entirely.</p>`
            },
            element: ".red-ui-sidebar-tab-bar > div",
            interactive: false
        },
        {
            title: {
                "en-US": "Dark theme unlocked"
            },
            description: {
                "en-US": `<p>Dark theme is now available with no need to enable it in settings.</p>
                          <p>This option may not be available if you have already installed a custom theme plugin.</p>`
            },
            element () {
                const themeSelector = $("#user-settings-view-dark-theme");
                if (themeSelector.length === 0) {
                    return $("#red-ui-tab-red-ui-settings-tab-view")
                }
                return themeSelector
            },
            interactive: false,
            prepare (done) {
                RED.actions.invoke('core:show-user-settings')
                 setTimeout(done,300);
            },
            complete () {
                $("#node-dialog-ok").trigger("click")
            }

        },
        {
            title: {
                "en-US": "Search toolbar"
            },
            description: {
                "en-US": `<p>Search your workspace to find nodes and flows quickly without disrupting your workflow.</p>`
            },
            element: "#red-ui-view-searchtools-search",
            interactive: false,
            image: 'images/editor-search-toolbar.png',
        }
        ,
        {
            title: {
                "en-US": "Pausing debug"
            },
            description: {
                "en-US": `<p>Pause the Debug sidebar's output when a busy flow is flooding you with messages.</p>`
            },
            prepare () {
                RED.actions.invoke("core:show-debug-tab")
            },
            element: "#red-ui-sidebar-debug-pause",
            interactive: false
        },
        {
            titleIcon: "fa fa-check",
            title: {
                "en-US": "You're all set"
            },
            description: {
                "en-US": `<p style="text-align: center">Now go build something amazing!</p>`
            }
        }
    ]
}
